### DISCUSSION TITLE
Feature Request: Database Obfuscation Tool

### DISCUSSION THREAD

**Discussion by hawkfish** [duckdb/duckdb#14215] [Ideas]
Why do you want this feature?User data is often highly confidential, which makes using it for debugging difficult.
At times, it may not even be possible to share data outside of the generating organisation,
so finding a way to obfuscate the data would be helpful in many such cases.

Obfuscation is not a simple problem because the issue under analysis may depend 
on some of the statistical properties of the data.
We can make a best effort in this regard, but ultimately the user must validate the output
of any tool we provide.

## Outline 

The proposed tool is actually two tools: one to obfuscate the data and another to rewrite any queries.
We will call the tools `obfuscate` and `rewrite` in what follows.

The `obfuscate` tool would work at the database file level and produce a second database with
remapped identifiers and strings, along with replacing the dependent variables ("measures")
with random values that have similar statistical properties.

The `obfuscate` tool would also produce a map for strings and identifiers 
that describes how to translate from the old values to the new ones.
This map can then be used by the `rewrite` tool to translate any query against the old database.

For the first version, we should target a configuration-free experience.
But as discussed below, there are a number of situations where it may be difficult
if not impossible to fully automate the obfuscation process.
For later versions, we could support reading a configuration specification 
to override our default behaviours.

## Obfuscation

There are several broad categories of data types, and each should be handled separately.
Specifically, these categories are 
metadata, dimensions (independent variables) and measures (dependent variables).
While some classifications are obvious 
(e.g., strings and temporal data are almost always dimensional)
some are hard to determine and may eventually require user overrides.

### Metadata

All field, table and schema names will be converted to new identifiers to reduce semantic leakage.
A common approach here is to replace the identifiers with a hash of the old string.
While this is straightforward, it makes reading the resulting queries difficult for humans.

An alternate approach is to use a two-state model based on the statistics of a natural language
(e.g., English).
This would generate identifiers that have a larger alphabet, more length variation,
and can usually be pronounced during discussions of the issue(!)

### Measures

Also known as dependent variables, these fields are usually numeric and aggregated.
For the purposes of debugging, we can generally replace them with random values.
To support filters transparently, these values should probably be constrained to the
original domain [min, max]. 

If a non-uniform distribution is required, we could potentially use the statistics
computed by `SUMMARIZE` to further shape the distribution,
but this runs the risk of semantic leakage.

### String Dimensions

Strings are similar to identifiers, but we should take care to model string length.
DuckDB strings can be inlined if they are 12 characters or less, which impacts memory footprint. 
Generated strings should probably follow this pattern if possible.
In particular, concatenation can be used to lengthen strings, 
and strings that are too long can be rejected and regenerated.

Because the nesting operations can convert metadata to strings, 
we need to share the string mapping between metadata and strings.

We may run into situations where strings are being cast to other types;
this situation would need to be handled by an override mechanism.
The sort order of strings may also turn out to be important.
To handle this, we would likely want to replace the start of the string with a value that enforces the order,
but this is an additional level of complexity that would need to be handled by an override mechanism.

### Numeric Dimensions

Besides strings, there are a number of data types that are typically used as independent variables.
Such variables are often used in joins and aggregation, so maintaining their statistics
and values is potentially important for debugging.
This suggests that these values should be maintained, 
and obfuscation restricted to the identifier name.

In addition to temporal types, small integer values are often used as join keys,
giving them dimensional semantics.
Deciding what constitutes "small" can be domain-dependent,
so eventually this may need to be configurable at the field level.
For the first version, we may be able to assume that 
integers narrower than `BIGINT` are dimensional, and wider types are measures.

## Rewrites

The `rewrite` tool will take a file containing queries 
and use the mapping file generated by `obfuscate` 
to replace identifiers and strings in the user's query with the mapped values.
If the bounds of scalar fields remain unchanged,
then most filters should work without modification.

### Identifiers

Rewriting string constants is straightforward because they are always quoted,
but identifiers are not required to be quoted.
This is not typically a problem (although out substitutions should always be quoted just in case)
but there may be situations where a keyword is allowed as an identifier.
If we wish to handle this situation, we would need to modify the AST of
the user's query and then convert it back to SQL.
For the first version, we should probably ignore this situation and ask the
user to quote such identifiers.

## User Validation

Once a user has generated an obfuscated database file, 
they can check that the rewritten queries reproduce their issue.
Moreover, they can examine the database to make sure it does not leak valuable semantic information.

**Comment by jimdickson007** [comment]
I am reasonably familiar with data security mechanism in other database(s) and consider obfuscation a tool in that arena.
In those databases userA sets data security privilege for userB.
DuckDB is single-user.
Am I mis-understanding your intention ?

**Reply by Tishj** [reply]
This is more related to the field of testing, a lot of users have sensitive datasets that they don't want to share, but when they run into bugs in the database, we often can't do much without a reproduction

**Reply by jimdickson007** [reply]
ah - got it, thanks for clarification
